include/quicly/cc.h

```diff
@@ -85,23 +148,63 @@ typedef struct st_quicly_cc_t {
     uint32_t num_loss_episodes;
 } quicly_cc_t;
 
+struct st_quicly_cc_impl_t {
+    void (*quicly_cc_init)(quicly_cc_t *, const quicly_cc_conf_t *, uint32_t);
+    void (*quicly_cc_on_acked)(quicly_cc_t *, const quicly_loss_t *, uint32_t, uint64_t, uint32_t, uint32_t);
+    void (*quicly_cc_on_lost)(quicly_cc_t *, const quicly_loss_t *, uint32_t, uint64_t, uint64_t, uint32_t);
+    void (*quicly_cc_on_persistent_congestion)(quicly_cc_t *, const quicly_loss_t *);
+};
+
+extern const struct st_quicly_cc_impl_t quicly_cc_reno_impl;
+extern const struct st_quicly_cc_impl_t quicly_cc_cubic_impl;
+
 /**
  * Initializes the congestion controller.
  */
-void quicly_cc_init(quicly_cc_t *cc, uint32_t initcwnd);
+static inline void quicly_cc_init(quicly_cc_t *cc, const quicly_cc_conf_t *conf, uint32_t initcwnd)
+{
+    memset(cc, 0, sizeof(quicly_cc_t));
+    cc->type = conf->type;
+
+    switch (cc->type) {
+    case CC_CUBIC:
+        cc->impl = &quicly_cc_cubic_impl;
+        break;
+    case CC_RENO_MODIFIED:
+    default:
+        cc->impl = &quicly_cc_reno_impl;
+        break;
+    }
+    cc->impl->quicly_cc_init(cc, conf, initcwnd);
+}
+
 /**
  * Called when a packet is newly acknowledged.
  */
-void quicly_cc_on_acked(quicly_cc_t *cc, uint32_t bytes, uint64_t largest_acked, uint32_t inflight, uint32_t max_udp_payload_size);
+static inline void quicly_cc_on_acked(quicly_cc_t *cc, const quicly_loss_t *loss, uint32_t bytes, uint64_t largest_acked,
```

Comment:
Inlined all of these since they only forward calls anyway.

Reply:
For simplicity, maybe call cc->impl->xxx directly from lib/quicly.c?

Reply:
Agreed - this was a leftover from an old version to avoid changes outside the cc source files. But once I had to add new parameters, that became necessary anyway.

Reply:
Left `quicly_cc_init` as is for now since there is more logic in there.

Reply:
Init function is also gone now due to new config interface.